File: Client/logger/logger.ts
/**
 * Centralized Logger with Category-Based Filtering
 * Supports runtime config reloading and PII scrubbing
 *
 * Adding New Log Categories:
 * 1. Add the category to logger.config.json (e.g., "search": true)
 *    - Categories can be added to config JSON even before code exists
 *    - The logger will handle unknown categories gracefully
 * 2. When implementing logging code:
 *    - Add to categories.ts: LogCategory type and LOG_CATEGORIES object
 *    - Add mapping in categoryToConfigKey function (categories.ts)
 *    - Add to LoggerConfig interface in logger.ts (optional, for type safety)
 * 3. Use the category in code: log.info(LOG_CATEGORIES.SEARCH, "message", data)
 *
 * The logger supports both defined categories (from categories.ts) and future
 * categories (from config JSON) - unknown categories are converted to camelCase
 * and checked against the config.
 */

import type { LogCategory, ApiSubcategory } from "./categories";
import {
  LOG_CATEGORIES,
  categoryToConfigKey,
  isAlwaysEnabled,
  apiSubcategoryToConfigKey,
} from "./categories";
import {
  scrubPII,
  maskSensitiveData,
  createSafeLogObject,
} from "./pii";

type LogLevel = "DEBUG" | "INFO" | "WARN" | "ERROR";

export interface ApiSubcategoryConfig {
  initialLoad: boolean;
  polling: boolean;
  pageMount: boolean;
  other: boolean;
}

export interface LoggerConfig {
  polling: boolean;
  pages: boolean;
  hooks: boolean;
  auth: boolean;
  http: boolean;
  api: boolean | ApiSubcategoryConfig;
  errors: boolean;
  security: boolean;
  search?: boolean;
  negotiation?: boolean;
  checklists?: boolean;
  calendar?: boolean;
  dashboard?: boolean;
  messages?: boolean;
  logLevel: LogLevel;
  // Allow additional category flags for future categories
  [key: string]: boolean | LogLevel | ApiSubcategoryConfig | undefined;
}

const LOG_LEVELS: Record<LogLevel, number> = {
  DEBUG: 0,
  INFO: 1,
  WARN: 2,
  ERROR: 3,
};

class Logger {
  private config: LoggerConfig;
  private isProcessing: boolean = false;
  private originalConsole: {
    log: typeof console.log;
    info: typeof console.info;
    warn: typeof console.warn;
    error: typeof console.error;
    debug: typeof console.debug;
  };

  constructor() {
    // Store original console methods
    this.originalConsole = {
      log: console.log.bind(console),
      info: console.info.bind(console),
      warn: console.warn.bind(console),
      error: console.error.bind(console),
      debug: console.debug.bind(console),
    };

    // Load initial config
    this.config = this.loadConfig();
  }

  /**
   * Load config from JSON file
   * In browser environment, this will need to be loaded via fetch
   * For now, we'll use a default config and allow runtime updates
   */
  private loadConfig(): LoggerConfig {
    // Default config
    const defaultConfig: LoggerConfig = {
      polling: true,
      pages: true,
      hooks: true,
      auth: true,
      http: true,
      api: {
        initialLoad: true,
        polling: true,
        pageMount: true,
        other: true,
      },
      errors: true,
      security: true,
      logLevel: "DEBUG",
    };

    // Try to load from config file (async, but we'll handle it)
    // For runtime reloading, we'll provide a method to update config
    return defaultConfig;
  }

  /**
   * Reload config from file (async)
   */
  async reloadConfig(): Promise<void> {
    try {
      const response = await fetch("/logger/logger.config.json");
      if (response.ok) {
        const config = await response.json();
        this.config = { ...this.config, ...config };
      }
    } catch (error) {
      // If config file doesn't exist or can't be loaded, use defaults
      this.originalConsole.warn(
        "[Logger] Failed to reload config, using current config",
        error,
      );
    }
  }

  /**
   * Update config programmatically (for runtime toggling)
   */
  updateConfig(updates: Partial<LoggerConfig>): void {
    this.config = { ...this.config, ...updates };
  }

  /**
   * Get current config (read-only)
   */
  getConfig(): Readonly<LoggerConfig> {
    return { ...this.config };
  }

  /**
   * Check if a category is enabled
   * Handles both defined categories (from categories.ts) and future categories (from config JSON)
   * Supports API subcategories when category is "API"
   */
  private isCategoryEnabled(
    category: LogCategory | string,
    subcategory?: ApiSubcategory,
  ): boolean {
    // Safety check: handle undefined/null categories
    if (!category || typeof category !== "string") {
      return false;
    }

    // Check if it's a known category that's always enabled
    if (category in LOG_CATEGORIES) {
      const logCategory = category as LogCategory;
      if (isAlwaysEnabled(logCategory)) {
        return true;
      }
    }

    // Try to get config key from category mapping (for defined categories)
    let configKey: string;
    try {
      if (category in LOG_CATEGORIES) {
        configKey = categoryToConfigKey(category as LogCategory);
      } else {
        // For future categories not yet in categories.ts, convert to camelCase
        // e.g., "SEARCH" -> "search"
        configKey = category
          .toLowerCase()
          .split("_")
          .map((word, index) =>
            index === 0
              ? word
              : word.charAt(0).toUpperCase() + word.slice(1),
          )
          .join("");
      }
    } catch {
      // If categoryToConfigKey fails, try direct camelCase conversion
      configKey = category
        .toLowerCase()
        .split("_")
        .map((word, index) =>
          index === 0 ? word : word.charAt(0).toUpperCase() + word.slice(1),
        )
        .join("");
    }

    // Check config
    const configValue = this.config[configKey as keyof LoggerConfig];

    // Handle API subcategories
    if (category === "API" || configKey === "api") {
      // If API config is an object (subcategories), check the subcategory
      if (typeof configValue === "object" && configValue !== null) {
        const apiConfig = configValue as ApiSubcategoryConfig;
        if (subcategory) {
          const subcategoryKey = apiSubcategoryToConfigKey(subcategory);
          return apiConfig[subcategoryKey as keyof ApiSubcategoryConfig] === true;
        }
        // If no subcategory specified, check if any subcategory is enabled
        return (
          apiConfig.initialLoad === true ||
          apiConfig.polling === true ||
          apiConfig.pageMount === true ||
          apiConfig.other === true
        );
      }
      // Backward compatibility: if API config is a boolean, use it directly
      return configValue === true;
    }

    // For other categories, check boolean value - default to false if not found
    return configValue === true;
  }

  /**
   * Check if log level is enabled
   */
  private isLevelEnabled(level: LogLevel): boolean {
    const currentLevel = LOG_LEVELS[this.config.logLevel];
    const messageLevel = LOG_LEVELS[level];
    return messageLevel >= currentLevel;
  }

  /**
   * Format log message with timestamp and category
   */
  private formatMessage(
    level: LogLevel,
    category: LogCategory | string,
    message: string,
    data?: unknown,
  ): string {
    if (this.isProcessing) {
      const timestamp = new Date().toISOString();
      return `[${timestamp}] [${level}] [${category}] ${message} [RECURSION_PREVENTED]`;
    }

    try {
      this.isProcessing = true;
      const timestamp = new Date().toISOString();
      const prefix = `[${timestamp}] [${level}] [${category}]`;

      if (data) {
        const scrubbedData = createSafeLogObject(data);
        return `${prefix} ${maskSensitiveData(message)} ${JSON.stringify(scrubbedData)}`;
      }

      return `${prefix} ${maskSensitiveData(message)}`;
    } catch {
      const timestamp = new Date().toISOString();
      return `[${timestamp}] [${level}] [${category}] ${message} [FORMAT_ERROR]`;
    } finally {
      this.isProcessing = false;
    }
  }

  /**
   * Debug logging
   */
  debug(
    category: LogCategory,
    message: string,
    data?: unknown,
    subcategory?: ApiSubcategory,
  ): void {
    if (
      !this.isCategoryEnabled(category, subcategory) ||
      !this.isLevelEnabled("DEBUG")
    ) {
      return;
    }

    try {
      const categoryLabel =
        subcategory && category === "API"
          ? `${category}:${subcategory}`
          : category;
      const formatted = this.formatMessage("DEBUG", categoryLabel, message, data);
      this.originalConsole.debug(formatted);
    } catch (error) {
      this.originalConsole.error("[Logger] Debug error:", error);
    }
  }

  /**
   * Info logging
   */
  info(
    category: LogCategory,
    message: string,
    data?: unknown,
    subcategory?: ApiSubcategory,
  ): void {
    if (
      !this.isCategoryEnabled(category, subcategory) ||
      !this.isLevelEnabled("INFO")
    ) {
      return;
    }

    try {
      const categoryLabel =
        subcategory && category === "API"
          ? `${category}:${subcategory}`
          : category;
      const formatted = this.formatMessage("INFO", categoryLabel, message, data);
      this.originalConsole.info(formatted);
    } catch (error) {
      this.originalConsole.error("[Logger] Info error:", error);
    }
  }

  /**
   * Warning logging
   */
  warn(
    category: LogCategory,
    message: string,
    data?: unknown,
    subcategory?: ApiSubcategory,
  ): void {
    if (
      !this.isCategoryEnabled(category, subcategory) ||
      !this.isLevelEnabled("WARN")
    ) {
      return;
    }

    try {
      const categoryLabel =
        subcategory && category === "API"
          ? `${category}:${subcategory}`
          : category;
      const formatted = this.formatMessage("WARN", categoryLabel, message, data);
      this.originalConsole.warn(formatted);
    } catch (error) {
      this.originalConsole.error("[Logger] Warn error:", error);
    }
  }

  /**
   * Error logging
   */
  error(
    category: LogCategory,
    message: string,
    error?: unknown,
    subcategory?: ApiSubcategory,
  ): void {
    if (
      !this.isCategoryEnabled(category, subcategory) ||
      !this.isLevelEnabled("ERROR")
    ) {
      return;
    }

    try {
      let errorData = error;

      // Handle Error objects
      if (error instanceof Error) {
        errorData = {
          name: error.name,
          message: error.message,
          stack: error.stack,
        };
      }

      const categoryLabel =
        subcategory && category === "API"
          ? `${category}:${subcategory}`
          : category;
      const formatted = this.formatMessage(
        "ERROR",
        categoryLabel,
        message,
        errorData,
      );
      this.originalConsole.error(formatted);
    } catch (err) {
      this.originalConsole.error("[Logger] Error logging error:", err);
    }
  }

  /**
   * Security event logging (always logs)
   */
  security(
    category: LogCategory,
    event: string,
    data?: unknown,
    subcategory?: ApiSubcategory,
  ): void {
    try {
      const scrubbedData = data ? createSafeLogObject(data) : undefined;
      const categoryLabel =
        subcategory && category === "API"
          ? `${category}:${subcategory}`
          : category;
      const formatted = this.formatMessage(
        "WARN",
        categoryLabel,
        `🔒 ${event}`,
        scrubbedData,
      );
      this.originalConsole.warn(formatted);
    } catch (error) {
      this.originalConsole.error("[Logger] Security logging error:", error);
    }
  }
}

// Export singleton instance
export const logger = new Logger();

// Convenience exports
export const log = {
  debug: (
    category: LogCategory,
    message: string,
    data?: unknown,
    subcategory?: ApiSubcategory,
  ) => logger.debug(category, message, data, subcategory),
  info: (
    category: LogCategory,
    message: string,
    data?: unknown,
    subcategory?: ApiSubcategory,
  ) => logger.info(category, message, data, subcategory),
  warn: (
    category: LogCategory,
    message: string,
    data?: unknown,
    subcategory?: ApiSubcategory,
  ) => logger.warn(category, message, data, subcategory),
  error: (
    category: LogCategory,
    message: string,
    error?: unknown,
    subcategory?: ApiSubcategory,
  ) => logger.error(category, message, error, subcategory),
  security: (
    category: LogCategory,
    event: string,
    data?: unknown,
    subcategory?: ApiSubcategory,
  ) => logger.security(category, event, data, subcategory),
  reloadConfig: () => logger.reloadConfig(),
  updateConfig: (updates: Partial<LoggerConfig>) =>
    logger.updateConfig(updates),
  getConfig: () => logger.getConfig(),
};

File: Server/logger/core/logger.py
"""
Centralized Logger with Category-Based Filtering
Supports runtime config reloading and PII scrubbing
"""

import json
import logging
import os
from datetime import datetime, timezone
from typing import Any

from ..config.config_model import LoggerConfig, LogLevel
from ..config.logger_env import should_export_logs_to_posthog
from ..config.resolve_logger_config import merge_logger_config_update, resolve_logger_config
from ..export.posthog_otlp import emit_structured_log
from ..security.pii import create_safe_log_object, mask_sensitive_data
from .categories import LogCategory
from .check_category_enabled import should_emit_log
from .parse_log_path import parse_log_path


class Logger:
    """Centralized logger with category-based filtering and PII scrubbing"""

    def __init__(self, config_path: str | None = None):
        """Initialize logger (config_path retained for tests that inject JSON overrides)."""
        self.config_path = config_path
        self.config = self._load_config()
        self.is_processing = False
        self._setup_python_logging()

    def _setup_python_logging(self) -> None:
        """Set up Python logging module"""
        # Get or create logger
        self._py_logger = logging.getLogger("app.logger")

        # Set level based on config
        level_map = {
            "DEBUG": logging.DEBUG,
            "INFO": logging.INFO,
            "WARN": logging.WARNING,
            "ERROR": logging.ERROR,
        }
        self._py_logger.setLevel(level_map.get(self.config.logLevel, logging.DEBUG))

        # Add handler if none exists
        if not self._py_logger.handlers:
            handler = logging.StreamHandler()
            handler.setFormatter(
                logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
            )
            self._py_logger.addHandler(handler)
        # Root also has a StreamHandler from configure_app_logging; avoid duplicate lines.
        self._py_logger.propagate = False

    def _load_config(self) -> LoggerConfig:
        """Load config with environment-aware defaults and optional test overrides."""
        overrides: dict[str, Any] = {}
        if self.config_path:
            try:
                if os.path.exists(self.config_path):
                    with open(self.config_path) as f:
                        overrides = json.load(f)
            except Exception:
                pass

        return resolve_logger_config(overrides or None)

    def reload_config(self) -> None:
        """Reload config from defaults (and optional test config_path)."""
        try:
            self.config = self._load_config()
            self._setup_python_logging()
        except Exception as e:
            self._py_logger.warning(f"[Logger] Failed to reload config: {e}")

    def update_config(self, updates: dict[str, Any]) -> None:
        """
        Update config programmatically (for runtime toggling)

        Args:
            updates: Dictionary of config updates
        """
        self.config = merge_logger_config_update(self.config, updates)
        self._setup_python_logging()

    def get_config(self) -> dict[str, Any]:
        """
        Get current config (read-only copy)

        Returns:
            Dictionary copy of current config
        """
        return self.config.to_dict()

    def _normalize_error_data(self, error: Exception | Any | None) -> Any | None:
        if error is None:
            return None
        if isinstance(error, Exception):
            error_data: dict[str, Any] = {
                "name": type(error).__name__,
                "message": str(error),
                "stack": None,
            }
            import traceback

            try:
                error_data["stack"] = traceback.format_exc()
            except Exception:
                pass
            return error_data
        return error

    def _scrubbed_payload(self, message: str, data: Any | None) -> tuple[str, Any | None]:
        safe_message = mask_sensitive_data(message)
        scrubbed_data = create_safe_log_object(data) if data is not None else None
        return safe_message, scrubbed_data

    def _emit(
        self,
        level: LogLevel,
        category_or_path: LogCategory | str,
        message: str,
        data: Any | None = None,
    ) -> None:
        """Emit to stdout and PostHog when category, level, and environment allow."""
        parsed = parse_log_path(category_or_path)
        if not should_emit_log(self.config, level, parsed.category):
            return

        try:
            safe_message, scrubbed_data = self._scrubbed_payload(message, data)
            formatted = self._format_message(
                level, parsed.category_label, safe_message, scrubbed_data
            )

            if should_export_logs_to_posthog():
                emit_structured_log(level, parsed.category_label, safe_message, scrubbed_data)

            py_level = {
                "DEBUG": self._py_logger.debug,
                "INFO": self._py_logger.info,
                "WARN": self._py_logger.warning,
                "SECURITY": self._py_logger.warning,
                "ERROR": self._py_logger.error,
            }.get(level, self._py_logger.info)
            py_level(formatted)
        except Exception as e:
            self._py_logger.error(f"[Logger] Emit error ({level}): {e}")

    def debug(self, category: LogCategory | str, message: str, data: Any | None = None) -> None:
        self._emit("DEBUG", category, message, data)

    def info(self, category: LogCategory | str, message: str, data: Any | None = None) -> None:
        self._emit("INFO", category, message, data)

    def warn(self, category: LogCategory | str, message: str, data: Any | None = None) -> None:
        self._emit("WARN", category, message, data)

    def error(
        self, category: LogCategory | str, message: str, error: Exception | Any | None = None
    ) -> None:
        self._emit("ERROR", category, message, self._normalize_error_data(error))

    def security(self, category: LogCategory | str, event: str, data: Any | None = None) -> None:
        scrubbed_data = create_safe_log_object(data) if data else None
        self._emit("SECURITY", category, f"🔒 {event}", scrubbed_data)

    def _format_message(
        self,
        level: LogLevel,
        category_label: str,
        message: str,
        data: Any | None = None,
    ) -> str:
        """
        Format log message with timestamp and category

        Args:
            level: Log level
            category: Log category
            message: Log message (already scrubbed when called from _emit)
            data: Optional data to include (already scrubbed when called from _emit)

        Returns:
            Formatted log message
        """
        if self.is_processing:
            timestamp = datetime.now(timezone.utc).isoformat()
            return f"[{timestamp}] [{level}] [{category_label}] {message} [RECURSION_PREVENTED]"

        try:
            self.is_processing = True
            timestamp = datetime.now(timezone.utc).isoformat()
            prefix = f"[{timestamp}] [{level}] [{category_label}]"

            if data is not None:
                data_str = json.dumps(data, default=str)
                return f"{prefix} {message} {data_str}"

            return f"{prefix} {message}"
        except Exception:
            timestamp = datetime.now(timezone.utc).isoformat()
            return f"[{timestamp}] [{level}] [{category_label}] {message} [FORMAT_ERROR]"
        finally:
            self.is_processing = False


# Export singleton instance
_logger_instance: Logger | None = None


def get_logger() -> Logger:
    """Get or create singleton logger instance"""
    global _logger_instance
    if _logger_instance is None:
        _logger_instance = Logger()
    return _logger_instance


# Convenience exports
logger = get_logger()


class LogProxy:
    """Convenience proxy for logger instance"""

    def debug(self, category: LogCategory | str, message: str, data: Any | None = None) -> None:
        logger.debug(category, message, data)

    def info(self, category: LogCategory | str, message: str, data: Any | None = None) -> None:
        logger.info(category, message, data)

    def warn(self, category: LogCategory | str, message: str, data: Any | None = None) -> None:
        logger.warn(category, message, data)

    def error(
        self, category: LogCategory | str, message: str, error: Exception | Any | None = None
    ) -> None:
        logger.error(category, message, error)

    def security(self, category: LogCategory | str, event: str, data: Any | None = None) -> None:
        logger.security(category, event, data)

    def reload_config(self) -> None:
        logger.reload_config()

    def update_config(self, updates: dict[str, Any]) -> None:
        logger.update_config(updates)

    def get_config(self) -> dict[str, Any]:
        return logger.get_config()


log = LogProxy()
